components/eamxx/src/diagnostics/aerocom_cld.cpp

```diff
@@ -0,0 +1,214 @@
+  metadata["3"] = "cldfrac_liq";
+  metadata["4"] = "cdnc";
+  metadata["5"] = "eff_radius_qc";
+  metadata["6"] = "eff_radius_qi";
+  metadata["7"] = "cldfrac_tot";
+}
+
+void AeroComCld::compute_diagnostic_impl() {
+  /* The goal of this routine/impl is to calculate properties at cloud top
+   * based on the AeroCom recommendation. See reference for routine
+   * get_subcolumn_mask in rrtmpg, where equation 14 is used for the
+   * maximum-random overlap assumption for subcolumn generation. We use
+   * equation 13, the column counterpart. The order is reversed to calculate
+   * cloud-bottom properties as well.
+   */
+  using KT  = KokkosTypes<DefaultDevice>;
+  using MT  = typename KT::MemberType;
+  using ESU = ekat::ExeSpaceUtils<typename KT::ExeSpace>;
+
+  using PF = scream::PhysicsFunctions<DefaultDevice>;
+
+  // Get the input fields
+  const auto tmid = get_field_in("T_mid").get_view<const Real **>();
+  const auto pden = get_field_in("pseudo_density").get_view<const Real **>();
+  const auto pmid = get_field_in("p_mid").get_view<const Real **>();
+  const auto qv   = get_field_in("qv").get_view<const Real **>();
+  const auto qc   = get_field_in("qc").get_view<const Real **>();
+  const auto qi   = get_field_in("qi").get_view<const Real **>();
+  const auto rel  = get_field_in("eff_radius_qc").get_view<const Real **>();
+  const auto rei  = get_field_in("eff_radius_qi").get_view<const Real **>();
+  const auto cld  = get_field_in("cldfrac_tot").get_view<const Real **>();
+  const auto nc   = get_field_in("nc").get_view<const Real **>();
+
+  auto dz = m_dz.get_view<Real **>();
+
+  // Get gravity acceleration constant from constants
+  using physconst = scream::physics::Constants<Real>;
+  // TODO: move tunable constant to namelist
+  constexpr auto q_threshold = 0.0;
+  // TODO: move tunable constant to namelist
+  constexpr auto cldfrac_tot_threshold = 0.001;
+
+  const auto policy = ESU::get_default_team_policy(m_ncols, m_nlevs);
+
+  const auto out = m_diagnostic_output.get_view<Real **>();
+
+  // zero out the outputs
+  Kokkos::deep_copy(out, 0.0);
+
+  const int nlevs = m_nlevs;
+  bool is_top     = (m_topbot == "Top");
+
+  Kokkos::parallel_for(
+      "Compute " + name(), policy, KOKKOS_LAMBDA(const MT &team) {
+        const int icol = team.league_rank();
+        // Subview the inputs at icol
+        auto tmid_icol = ekat::subview(tmid, icol);
+        auto pden_icol = ekat::subview(pden, icol);
+        auto pmid_icol = ekat::subview(pmid, icol);
+        auto qv_icol   = ekat::subview(qv, icol);
+        auto qc_icol   = ekat::subview(qc, icol);
+        auto qi_icol   = ekat::subview(qi, icol);
+        auto rel_icol  = ekat::subview(rel, icol);
+        auto rei_icol  = ekat::subview(rei, icol);
+        auto cld_icol  = ekat::subview(cld, icol);
+        auto nc_icol   = ekat::subview(nc, icol);
+
+        // We need dz too
+        // TODO: deduce dz_icol type, etc. on the spot here w/o field cloning?
```

Comment:
This comment is outdated?

Reply:
I'm not sure what you need to deduce. The comment says you want to deduce the type of dz_icol, but why? You know it's a 1d view, and that's all you need, right?

As for temp views: inside kernel, you can't declare managed views, you can only create unmanaged views of existing memory (like subviews). Outside a kernel, you can declare any view you like, though be aware that, again, allocating views at runtime is not very efficient; you should create any scratch view at init time (unless they are 'scratch' in the kokkos sense, meaning they use the `ScratchMemorySpace`, which is a per-kernel scratch memory; but we don't really use that in EAMxx (yet)).

Reply:
I mean, doesn't have to be a `Field` obj, you really only need a view. You could use an `ekat::WorkspaceManager`, to be a tad more efficient, if you wanted.